plugins/hypervisors/xenserver/src/com/cloud/hypervisor/xenserver/resource/CitrixResourceBase.java

```diff
@@ -1415,6 +1417,10 @@ public VM createWorkingVM(final Connection conn, final String vmName, final Stri
             vbdr.userdevice = "autodetect";
             vbdr.mode = Types.VbdMode.RW;
             vbdr.type = Types.VbdType.DISK;
+            Long deviceId = volumeTO.getDeviceId();
+            if (deviceId != null && (!isDeviceUsed(conn, vm, deviceId) || deviceId > 3)) {
```

Comment:
There is this one also https://github.com/apache/cloudstack/pull/1829.
However, after sorting out the tests failures and with the code approved I don't think that this would be a reason to not take the PR forward.